Enhance config usage by multi-node benchmarks.

cmccoy opened the pull request:
See #512.

* Standardizes names for VM groups in database and Hadoop benchmarks to
  `workers` (data serving / processing nodes) and `clients` (load generating
  nodes). Systems with a master have an additional `master` node.
* Splits out VM roles for CloudSuite benchmarks.
* Adds multi-server support to Aerospike benchmark.

cmccoy commented:
@ehankland: would you mind taking a look?

ehankland commented on perfkitbenchmarker/linux_benchmarks/cloudsuite_websearch_benchmark.py:
```diff
@@ -107,7 +110,8 @@ def _CheckHeapSize(heap_size_str):
   config = configs.LoadConfig(BENCHMARK_CONFIG, user_config, BENCHMARK_NAME)
   num_clients = FLAGS.cs_websearch_num_clients
-  config['vm_groups']['default']['vm_count'] = num_clients + NUM_SERVERS
+  config['vm_groups']['workers']['vm_count'] = NUM_SERVERS
```
Doesn't look like we need the "NUM_SERVERS" constant anymore now that there are separate groups.

ehankland commented on perfkitbenchmarker/linux_benchmarks/hadoop_terasort_benchmark.py:
```diff
@@ -54,7 +58,7 @@
   config = configs.LoadConfig(BENCHMARK_CONFIG, user_config, BENCHMARK_NAME)
   if FLAGS['num_vms'].present:
-    config['vm_groups']['default']['vm_count'] = FLAGS.num_vms
+    config['vm_groups']['workers']['vm_count'] = FLAGS.num_vms - 1
```
Based on the description I would expect this to just be FLAGS.num_vms not FLAGS.num_vms -1.

ehankland commented on perfkitbenchmarker/linux_benchmarks/mongodb_ycsb_benchmark.py:
```diff
@@ -40,15 +40,22 @@
       vm_spec: *default_single_core
       disk_spec: *default_500_gb
-      vm_count: 2
+      vm_count: 1
```
set it to null and get rid of lines 56+57 below

cmccoy replied:
Done.

ehankland commented on perfkitbenchmarker/linux_benchmarks/mongodb_ycsb_benchmark.py:
```diff
@@ -40,15 +40,22 @@
+      vm_count: 1
+    clients:
+      vm_spec: *default_single_core
+      vm_count: 1
```
Not needed since this is just getting overwritten below anyway

ehankland commented on perfkitbenchmarker/linux_benchmarks/redis_ycsb_benchmark.py:
```diff
@@ -33,9 +33,12 @@
+    workers:
       vm_spec: *default_single_core
-      vm_count: null
+      vm_count: 1
```
default vm_count is 1 so you don't need this

ehankland commented on perfkitbenchmarker/linux_benchmarks/redis_ycsb_benchmark.py:
```diff
@@ -33,9 +33,12 @@
+      vm_count: 1
+    clients:
+      vm_spec: *default_single_core
+      vm_count: 1
```
same as above

ehankland commented on perfkitbenchmarker/linux_benchmarks/aerospike_benchmark.py:
```diff
@@ -54,10 +54,13 @@
+      vm_count: null
+    client:
+      vm_spec: *default_single_core
+      vm_count: 1
```
dont need this

ehankland commented on perfkitbenchmarker/linux_benchmarks/cloudsuite_web_serving_benchmark.py:
```diff
@@ -55,16 +53,23 @@
       vm_spec: *default_single_core
       disk_spec: *default_500_gb
-      vm_count: 3
+      vm_count: 1
```
dont need any of these vm_counts

ehankland commented on perfkitbenchmarker/linux_benchmarks/hadoop_terasort_benchmark.py:
```diff
@@ -39,12 +39,16 @@
       vm_spec: *default_single_core
       disk_spec: *default_500_gb
-      vm_count: 9
+      vm_count: 1
```
same as the other comments about vm_count - dont need this

ehankland commented on perfkitbenchmarker/linux_benchmarks/cloudsuite_websearch_benchmark.py:
```diff
@@ -42,10 +42,13 @@
       vm_count: null
+    clients:
+      vm_spec: *default_single_core
+      vm_count: null
```
Don't really need this since it gets overridden below

ehankland commented:
LGTM - thanks for taking the time to do this

cmccoy commented:
Thanks. All done.

ehankland commented:
:+1:

cmccoy commented:
Re-tested everything and added `cassandra_ycsb`.  Both CloudSuite benchmarks failed for me - I've removed that change until I sort out what happened.

@ehankland: would you mind taking a quick look at `cassandra_stress` and `cassandra_ycsb`?

cmccoy commented:
Fixed cloudsuite as well.

ehankland commented:
LGTM - there are a few places where vm_count gets set unconditionally to some flag value (e.g. ycsb_client_vms) - it would be nice if we checked for the presence of the flag so that setting the vm_count in a user config would work.

cmccoy commented:
> there are a few places where vm_count gets set unconditionally to some flag value (e.g. ycsb_client_vms) - it would be nice if we checked for the presence of the flag so that setting the vm_count in a user config would work.

For the YCSB case, done in 61c674ce6a743f70c4e9a98f2e5d76a98865cdd4
